juniper/src/executor/look_ahead_lazy.rs

```diff
@@ -0,0 +1,1555 @@
+        self.children
+            .iter()
+            .find(|child| child.field_name() == name)
+            .cloned()
+    }
+
+    /// Checks if a child selection with the specified `name` exists.
+    ///
+    /// If a child has an alias, it will only match if the alias matches the specified `name`.
+    pub fn has_child(&self, name: &str) -> bool {
+        self.select(name).is_some()
+    }
+
+    /// Returns the (possibly aliased) names of the top level children from the current selection.
+    pub fn names(&self) -> impl Iterator<Item = &'a str> + '_ {
+        self.children.iter().map(|selection| selection.field_name())
+    }
+
+    /// Iterate over the children, by reference.
+    pub fn iter(&self) -> impl Iterator<Item = &LookAheadSelection<'a, S>> + '_ {
+        self.children.iter()
+    }
+}
+
+impl<'a, S: ScalarValue> IntoIterator for LookAheadChildren<'a, S> {
+    type Item = LookAheadSelection<'a, S>;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.children.into_iter()
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(super) enum SelectionSource<'a, S: ScalarValue> {
+    Field(&'a Field<'a, S>),
+    Spread {
+        field_name: &'a str,
+        set: Option<&'a [Selection<'a, S>]>,
+    },
+}
+
+/// A selection performed by a query
+#[derive(Clone, Copy, Debug)]
+pub struct LookAheadSelection<'a, S: ScalarValue + 'a> {
+    source: SelectionSource<'a, S>,
+    applies_for: Applies<'a>,
+    vars: &'a Variables<S>,
+    fragments: &'a HashMap<&'a str, Fragment<'a, S>>,
+}
+
+impl<'a, S: ScalarValue> LookAheadSelection<'a, S> {
+    pub(super) fn new(
+        source: SelectionSource<'a, S>,
+        vars: &'a Variables<S>,
+        fragments: &'a HashMap<&'a str, Fragment<'a, S>>,
+    ) -> Self {
+        Self {
+            source,
+            applies_for: Applies::All,
+            vars,
+            fragments,
+        }
+    }
+
+    /// Returns the original name of the field, represented by the current selection.
+    pub fn field_original_name(&self) -> &'a str {
+        match self.source {
+            SelectionSource::Field(field) => field.name.item,
+            SelectionSource::Spread { field_name, .. } => field_name,
+        }
+    }
+
+    /// Returns the alias of the field, represented by the current selection, if any.
+    pub fn field_alias(&self) -> Option<&'a str> {
+        match self.source {
+            SelectionSource::Field(field) => field.alias.map(|alias| alias.item),
+            SelectionSource::Spread { .. } => None,
+        }
+    }
+
+    /// Returns the (potentially aliased) name of the field, represented by the current selection.
+    pub fn field_name(&self) -> &'a str {
+        self.field_alias()
+            .unwrap_or_else(|| self.field_original_name())
+    }
+
+    /// Returns the top level arguments from the current selection.
+    pub fn arguments(&self) -> Option<LookAheadArguments<'a, S>> {
```

Comment:
This now encodes whether arguments are present or not, but I'm not sure whether this is necessary. The point was to meaningfully replace `fn LookAheadMethods::has_arguments`.

Reply:
This could change into ` -> impl Iterator<Item = LookAheadArgument>`. But then does `has_arguments` need to be added back? One could achieve the same with ortogonal features  `look_ahead.arguments().next().is_some()`.

Reply:
I agree, I'll reintroduce `has_arguments` and do `arguments() -> LookAheadArguments`.